lib/image_optim/cmd.rb

```diff
@@ -15,6 +18,44 @@ def run(*args)
         success
       end
 
+      def support_timeout?
+        if defined?(JRUBY_VERSION)
+          JRUBY_VERSION >= '9.0.0.0'
+        else
+          RUBY_VERSION >= '1.9'
+        end
+      end
+
+      # Run the specified command, and kill it off if it runs longer
+      # than `timeout` seconds.
+      #
+      # Return success status
+      # Will raise an error when command timeouts
+      def run_with_timeout(timeout, *args)
+        return run(*args) unless timeout > 0 && support_timeout?
+
+        success = false
+        init_options!(args)
+
+        begin
+          stdin, stdout, thread = Open3.popen2(*args)
+          stdin.close
+          stdout.close
+
+          pid = thread[:pid]
+
+          if thread.join(timeout).nil?
+            cleanup_process(pid)
```

Comment:
`cleanup_process` creates a thread which is never joined, then the watching thread is killed, I think in this case it is better to wait for cleanup to finish